gtsfm/frontend/correspondence_generator/mast3r_correspondence_generator.py

```diff
@@ -0,0 +1,341 @@
+
+    Args:
+        image: The input image to be preprocessed.
+        img_id: An identifier for the image.
+        device: The device (e.g., 'cpu' or 'cuda') to which the image tensor should be moved.
+
+    Returns:
+        A tuple containing:
+        - A dictionary with the preprocessed image tensor, its true shape, id, and instance string.
+        - A tuple (x, y) representing the start coordinates of the crop applied to the image.
+        - The scaling factor applied to the image during resizing.
+    """
+    H1, _, _ = image.shape
+
+    # resize to longest edge being `size`.
+    img = resize_image(image, long_edge_size=512).value_array
+    H, W, _ = img.shape
+    scale_factor = float(H) / H1
+
+    cx, cy = W // 2, H // 2
+    halfw, halfh = ((2 * cx) // GRID_DIVISOR) * PATCH_MULTIPLIER, ((2 * cy) // GRID_DIVISOR) * PATCH_MULTIPLIER
+    crop_start = (cx - halfw, cy - halfh)
+    img = img[crop_start[1] : cy + halfh, crop_start[0] : cx + halfw]
+    ImgNorm = tvf.Compose([tvf.ToTensor(), tvf.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+    return (
+        dict(
+            img=ImgNorm(img)[None].to(device),
+            true_shape=np.int32([img.shape[:2]]),
+            idx=img_id,
+            instance=str(img_id),
+        ),
+        crop_start,
+        scale_factor,
+    )
+
+
+def symmetric_inference(model, img1, img2, device):
+    """Performs symmetric inference using the MASt3R model on a pair of images.
+
+    Args:
+        model: The MASt3R model.
+        img1: Preprocessed dictionary for the first image.
+        img2: Preprocessed dictionary for the second image.
+        device: The device (e.g., 'cpu' or 'cuda') to run the inference on.
+
+    Returns:
+        A tuple of four results (res11, res21, res22, res12) from the symmetric decoding process.
+    """
+    shape1 = torch.from_numpy(img1["true_shape"]).to(device, non_blocking=True)
+    shape2 = torch.from_numpy(img2["true_shape"]).to(device, non_blocking=True)
+    img1 = img1["img"].to(device, non_blocking=True)
+    img2 = img2["img"].to(device, non_blocking=True)
+
+    # compute encoder only once
+    feat1, feat2, pos1, pos2 = model._encode_image_pairs(img1, img2, shape1, shape2)
+
+    def decoder(feat1, feat2, pos1, pos2, shape1, shape2):
+        dec1, dec2 = model._decoder(feat1, pos1, feat2, pos2)
+
+        with torch.cuda.amp.autocast(enabled=False):
+            res1 = model.downstream_head1([tok.float() for tok in dec1], shape1[0])
+            res2 = model.downstream_head2([tok.float() for tok in dec2], shape2[0])
+        return res1, res2
+
+    # decoder 1-2
+    res11, res21 = decoder(feat1, feat2, pos1, pos2, shape1, shape2)
+    # decoder 2-1
+    res22, res12 = decoder(feat2, feat1, pos2, pos1, shape2, shape1)
+
+    return (res11, res21, res22, res12)
+
+
+def extract_correspondences(feats, qonfs, subsample=8, device=None, ptmap_key="pred_desc"):
```

Comment:
Maybe this should be a static method of the class if no other classes use it?

Reply:
not necessarily. I basically copied these functions from mast3r codebase (so theyre not super readable). I can make them staticmethods in the gtsfm class, but I wanted to keep that class uncluttered. To some degree, readers dont need to fully understand what `extract_correspondences` does, but can still understand the class at a high level. But sure, it can be a staticmethod if you prefer that.

Reply:
I think I would prefer it to be a static class method since it's only valid for the Mast3rCorrespondenceGenerator class. I would also prefix it with an underscore to signify it's for internal use only.